addons/addon-base-raas/packages/base-raas-services/lib/data-egress/data-egress-service.js

```diff
@@ -124,15 +138,114 @@ class DataEgressService extends Service {
+        });
+    } catch (error) {
+      throw this.boom.notFound(`Error in terminating egress store: ${JSON.stringify(error)}`, true);
+    }
+    const egressStoreInfo = egressStoreScanResult[0];
+    const isEgressStoreOwner = egressStoreInfo.createdBy === curUser;
+    if (!isAdmin(requestContext) && !isEgressStoreOwner) {
+      throw this.boom.forbidden(
+        `You are not authorized to terminate the egress store. Please contact your administrator.`,
+        true,
+      );
+    }
+
+    const s3Service = await this.service('s3Service');
+    const egressStoreStatus = egressStoreInfo.status;
+
+    if (egressStoreStatus.toUpperCase() === PROCESSING_STATUS_CODE) {
+      throw this.boom.forbidden(
+        `Egress store: ${egressStoreInfo.id} is still in processing, workspce can not be terminated before egress request is processed.`,
+        true,
+      );
+    } else if (egressStoreStatus.toUpperCase() === PROCESSED_STATUS_CODE) {
+      // ONLY terminate the egress store if it has been processed
+
+      try {
+        await s3Service.clearPath(egressStoreInfo.s3BucketName, egressStoreInfo.s3BucketPath);
+      } catch (error) {
+        throw this.boom.badRequest(
+          `Error in deleting egress store:${egressStoreInfo.s3BucketName} in bucket: ${egressStoreInfo.s3BucketPath}`,
+          true,
+        );
+      }
+      egressStoreInfo.status = TERMINATED_STATUS_CODE;
+      egressStoreInfo.updatedBy = curUser;
+      egressStoreInfo.updatedAt = new Date().toISOString();
+      await runAndCatch(
+        async () => {
+          return this._updater()
+            .condition('attribute_exists(id)') // yes we need this to ensure the egress store does not exist already
+            .key({ id: egressStoreInfo.id })
+            .item(egressStoreInfo)
+            .update();
+        },
+        async () => {
+          throw this.boom.badRequest(`Egress Store with id "${egressStoreInfo.id}" got updating error`, true);
+        },
+      );
+
+      const egressStore = {
+        id: `egress-store-${environmentId}`,
+        readable: true,
+        writeable: true,
+        bucket: egressStoreInfo.s3BucketName,
+        prefix: egressStoreInfo.s3BucketPath,
+        envPermission: {
+          read: true,
+          write: true,
+        },
+        status: 'reachable',
+        createdBy: egressStoreInfo.createdBy,
+        workspaceId: environmentId,
+        projectId: egressStoreInfo.projectId,
+        resources: [
+          {
+            arn: `arn:aws:s3:::${egressStoreInfo.s3BucketName}/${environmentId}/`,
+          },
+        ],
+      };
+
+      // Remove egress store related s3 policy from the s3 bucket
+      const memberAccountId = await this.getMemberAccountId(requestContext, environmentId);
+
+      const lockService = await this.service('lockService');
+      const lockId = `bucket-policy-access-${egressStoreInfo.s3BucketName}`;
```

Comment:
Given that create/terminate are changing the same DDB row- should we run all create/terminate methods under an environmentId specific lock to avoid potential race conditions.

Reply:
Overall, I think the create/terminate methods won't run into race condition since they happens in fixed sequential. But I did applied the lock in the fix just in case the race condition when updating processing and processed status in other API calls. 
fix in commit: c33f4053406ffd30ff50181448e5cba7a7925a90